File: typescript/src/part1/chapter4/validated.ts
import { match, P } from "ts-pattern";
import { Either } from "src/part1/chapter4/either"

export type Validated<E, A> = Invalid<E> | Valid<A>
export interface Invalid<E> {
    type: 'Invalid'
    value: Array<E>
}
export interface Valid<A> {
    type: 'Valid'
    value: A
}

export const toEither = <E, A>(v: Validated<E, A>): Either<Array<E>, A> =>
    match(v)
        .with({ type: 'Invalid' }, (invalid) => ({ type: 'Left', value: invalid.value } as Either<Array<E>, A>))
        .with({ type: 'Valid' }, (valid) => ({ type: 'Right', value: valid.value } as Either<Array<E>, A>))
        .exhaustive()

export const map = <E, A, B>(v: Validated<E, A>, f: (a: A) => B): Validated<E, B> =>
    match(v)
        .with({ type: 'Invalid' }, (invalid) => ({ ...invalid }))
        .with({ type: 'Valid' }, (valid) => ({ ...valid, value: f(valid.value) }))
        .exhaustive()

export const map2 = <E, A, B, C>(a: Validated<E, A>, b: Validated<E, B>, f: (a: A, b: B) => C): Validated<E, C> => {
    if (a.type === 'Valid' && b.type === 'Valid') { return { type: 'Valid', value: f(a.value, b.value) } }
    else if (a.type === 'Invalid' && b.type === 'Valid') { return a }
    else if (a.type === 'Valid' && b.type === 'Invalid') { return b }
    else { return { type: 'Invalid', value: [...(a as Invalid<E>).value, ...(b as Invalid<E>).value] } }
}

export const fromEither = <E, A>(e: Either<Array<E>, A>): Validated<E, A> =>
    match(e)
        .with({ type: 'Left' }, (left) => ({ type: 'Invalid', value: left.value } as Validated<E, A>))
        .with({ type: 'Right' }, (right) => ({ type: 'Valid', value: right.value } as Validated<E, A>))
        .exhaustive()

export const traverse = <E, A, B>(as: Array<A>, f: (a: A) => Validated<E, B>): Validated<E, Array<B>> =>
    as.reduce((acc: Validated<E, Array<B>>, a: A) => map2(f(a), acc, (b: B, bs: Array<B>) => [b, ...bs]), { type: 'Valid', value: [] } as Validated<E, Array<B>>)

export const sequence = <E, A>(vs: Array<Validated<E, A>>): Validated<E, Array<A>> =>
    traverse(vs, (x) => x)

File: python/src/part1/chapter4/validated.py
from __future__ import annotations

from dataclasses import dataclass
from functools import reduce
from typing import Generic, TypeVar, final, TypeAlias, Callable, cast

from part1.chapter4.either import Either, Left, Right

E = TypeVar("E", covariant=True)
A = TypeVar("A", covariant=True)
B = TypeVar("B", covariant=True)
C = TypeVar("C", covariant=True)


@dataclass(frozen=True)
class Validated(Generic[E, A]):
    value: list[E] | A

    def map(self, f: Callable[[A], B]) -> Validated[E, B]:
        match self.pattern:
            case Invalid(_):
                return cast(Validated[E, B], self)
            case Valid(value):
                return cast(Validated[E, B], Valid(f(value)))

    def map2(self, other: Validated[E, B], f: Callable[[A, B], C]) -> Validated[E, C]:
        match (self.pattern, other.pattern):
            case (Valid(aa), Valid(bb)):
                return cast(Validated[E, C], Valid(f(aa, bb)))
            case (Invalid(es), Valid(_)):
                return cast(Validated[E, C], Invalid(es))
            case (Valid(_), Invalid(es)):
                return cast(Validated[E, C], Invalid(es))
            case (Invalid(es1), Invalid(es2)):
                return cast(Validated[E, C], Invalid(es1 + es2))
            case _:
                raise ValueError(self, other)

    @staticmethod
    def from_either(either: Either[list[E], A]) -> Validated[E, A]:
        match either.pattern:
            case Left(errors):
                return cast(Validated[E, A], Invalid(errors))
            case Right(value):
                return cast(Validated[E, A], Valid(value))

    def to_either(self) -> Either[list[E], A]:
        match self.pattern:
            case Invalid(error):
                return cast(Either[list[E], A], Left(error))
            case Valid(value):
                return cast(Either[list[E], A], Right(value))

    @staticmethod
    def traverse(xs: list[A], f: Callable[[A], Validated[E, B]]) -> Validated[E, list[B]]:
        return reduce(lambda acc, a: f(a).map2(acc, lambda b, bs: [b] + bs), xs, Valid[E, B]([]))

    @staticmethod
    def sequence(vs: list[Validated[E, A]]) -> Validated[E, list[A]]:
        return Validated.traverse(vs, lambda x: x)

    @property
    def pattern(self) -> SubType[E, A]:
        return cast(SubType[E, A], self)


@final
class Invalid(Validated[E, A]):
    value: list[E]


@final
class Valid(Validated[E, A]):
    value: A


SubType: TypeAlias = Invalid[E, A] | Valid[E, A]
